### PR TITLE
Review comments2

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- force implementation of `__eq__` for Events
- add a helper class for Events without attributes to avoid duplicating their `__eq__` implementation
- mark `Event` class as abstract

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/shared/core/events.py [hunk 1]
```diff
@@ -314,21 +317,30 @@ def apply_to(self, tracker: "DialogueStateTracker") -> None:
 
     def __eq__(self, other: Any) -> bool:
         """Compares object with other object."""
-        if not isinstance(other, self.__class__):
-            return NotImplemented
-
-        return True
+        # Every class should implement this
```
Comment (joejuzl) [inline rasa/shared/core/events.py:320]: Does this mean that `Event` should be an abstract class? (And this an abstract method?)

File: rasa/shared/core/events.py [hunk 2]
```diff
@@ -314,21 +317,30 @@ def apply_to(self, tracker: "DialogueStateTracker") -> None:
...
         """Returns text representation of event."""
         return f"{self.__class__.__name__}()"
 
 
-# noinspection PyProtectedMember
+class EventWithoutExtraAttributes(Event, ABC):
+    """Class to deduplicate common behavior for events without additional attributes."""
+
+    def __eq__(self, other: Any) -> bool:
+        """Returns text representation of event."""
```
Comment (joejuzl) [inline rasa/shared/core/events.py:332]: This comment seems wrong.

File: rasa/shared/core/events.py [hunk 3]
```diff
@@ -314,21 +317,30 @@ def apply_to(self, tracker: "DialogueStateTracker") -> None:
...
+        # Every class should implement this
+        raise NotImplementedError()
 
     def __str__(self) -> Text:
         """Returns text representation of event."""
         return f"{self.__class__.__name__}()"
 
 
-# noinspection PyProtectedMember
+class EventWithoutExtraAttributes(Event, ABC):
```
Comment (joejuzl) [inline rasa/shared/core/events.py:328]: I find this name is confusing and likely to form a grab bag. Maybe something along the lines of `AlwaysEqualEventMixin`?

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: rasa/shared/core/events.py (hunk 1 above)
Comment (wochinge) [reply rasa/shared/core/events.py:320]: good catch! Should have been an abstract class all along

File: rasa/shared/core/events.py (hunk 3 above)
Comment (wochinge) [reply rasa/shared/core/events.py:328]: Mhm, it's not always equal. It's only equal if it has the same type 🤔 In my opinion it'd actually be ok if it ends up as grab bag, e.g. I see that most of this classes have `as_story_string` implemented with `return self.type_name`

Comment (wochinge) [reply rasa/shared/core/events.py:328]: actually I like your name suggestion, will go with it 👍🏻

Comment (joejuzl) [reply rasa/shared/core/events.py]: I do see your point though! If we add more functionality to it we can change the name again to fit.

Comment (wochinge) [reply rasa/shared/core/events.py]: Yep 👍🏻 Let's keep it narrow for now. We can't fix all problems in the `e2e` branch anyway

File: rasa/shared/core/events.py [hunk 4]
```diff
@@ -314,21 +317,30 @@ def apply_to(self, tracker: "DialogueStateTracker") -> None:
 
     def __eq__(self, other: Any) -> bool:
```
Comment (joejuzl) [inline rasa/shared/core/events.py:318]: We can also make this an `abstractmethod` now right?

**Review by joejuzl — APPROVED** [review]
Looks good. Just one final point.

File: rasa/shared/core/events.py (hunk 4 above)
Comment (wochinge) [reply rasa/shared/core/events.py:321]: actually didn't know about this decorator. Added it 👍🏻

**Pull request merged**